File: design-patterns/ts/visitor/hosting-cost/file-storage-service.ts
// file-storage-service.ts

import HostingCalculatorVisitor from "./hosting-calculator-visitor";
import Service from "./service";


class FileStorageService implements Service {

    private readonly pricePerGB: number = 1.70;
    private quantity: number;

    constructor(quantity: number) {
        this.quantity = quantity;
    }

    getPricePerGB(): number {
        return this.pricePerGB;
    }

    getQuantity(): number {
        return this.quantity;
    }

    accept(hostingCalculatorVisitor: HostingCalculatorVisitor): number {
        return hostingCalculatorVisitor.visit(this);
    }
}

export default FileStorageService;

File: design-patterns/go/visitor/hosting_cost/file_storage_service.go
// file_storage_service.go

package main

type FileStorageService struct {
	pricePerGB float64
	quantity   int
}

func NewFileStorageService(quantity int) (fileStorageService *FileStorageService) {
	fileStorageService = &FileStorageService{}
	fileStorageService.pricePerGB = 1.70
	fileStorageService.quantity = quantity
	return
}

func (fileStorageService *FileStorageService) GetPricePerGB() float64 {
	return fileStorageService.pricePerGB
}

func (fileStorageService *FileStorageService) GetQuantity() int {
	return fileStorageService.quantity
}

func (fileStorageService *FileStorageService) Accept(hostingCalculatorVisitor HostingCalculatorVisitor) float64 {
	return hostingCalculatorVisitor.FileVisit(fileStorageService)
}
